File: benchmarks-ocaml/church_list_reverse_and_sum.ml
open Interpreter

let church_nil = Lambda (Lambda (tvar 0))

let church_cons =
  Lambda
    (Lambda
       (Lambda
          (Lambda
             (Apply
                (Apply (tvar 1, tvar 3), Apply (Apply (tvar 2, tvar 1), tvar 0))))))

let add = BinaryOp ( + )

let church_sum_list = Lambda (Apply (Apply (tvar 0, add), Const 0))

let church_reverse =
  Lambda
    (Apply
       ( Apply
           ( tvar 0,
             Lambda
               (Lambda
                  (Lambda
                     (Lambda
                        (Apply
                           ( Apply (tvar 2, tvar 1),
                             Apply (Apply (tvar 1, tvar 3), tvar 0) ))))) ),
         church_nil ))

let rec generate_list n =
  let rec go i acc =
    if i = 0 then acc
    else go (i - 1) (Apply (Apply (church_cons, Const (i - 1)), acc))
  in
  go n church_nil

let benchmark_term =
  Apply (church_sum_list, Apply (church_reverse, generate_list 1000000))

let () =
  match denote [] benchmark_term with
  | VConst n -> Printf.printf "%d\n" n
  | _ -> failwith "Expected a constant result!"

File: benchmarks-haskell/church-list-reverse-and-sum.hs
import Interpreter

churchNil :: Term
churchNil = Lambda (Lambda (TVar 0))

churchCons :: Term
churchCons =
  Lambda
    ( Lambda
        ( Lambda
            (Lambda (Apply (Apply (TVar 1) (TVar 3)) (Apply (Apply (TVar 2) (TVar 1)) (TVar 0))))
        )
    )

add :: Term
add = BinaryOp (+)

churchSumList :: Term
churchSumList =
  Lambda
    ( Apply (Apply (TVar 0) add) (Const 0)
    )

churchReverse :: Term
churchReverse =
  Lambda
    ( Apply
        ( Apply
            (TVar 0)
            ( Lambda
                ( Lambda
                    ( Lambda
                        ( Lambda
                            ( Apply
                                (Apply (TVar 2) (TVar 1))
                                (Apply (Apply (TVar 1) (TVar 3)) (TVar 0))
                            )
                        )
                    )
                )
            )
        )
        churchNil
    )

generateList :: Int -> Term
generateList n = go n churchNil
  where
    go 0 acc = acc
    go i acc = go (i - 1) (Apply (Apply churchCons (Const (i - 1))) acc)

benchmarkTerm :: Term
benchmarkTerm = Apply churchSumList (Apply churchReverse (generateList 1000000))

main :: IO ()
main = case denote [] benchmarkTerm of
  VConst n -> print n
  _ -> error "Expected a constant result!"
